### ISSUE TITLE
Enable `noExplicitAny` in `./tsconfig.json`.

### ISSUE THREAD

**Issue by lillallol** [microsoft/TypeScript#54933]
## ⭐ Suggestion & 📃 Motivating Example & 💻  Use Cases

The following `.js` code, in project with a `./tsconfig.json` that has `allowJs:true`, `checkJs:true`,`noImplicitAny:true`, does not lint error:

```js
const mySet = new Set;
const myMap = new Map;
const myWeakMap = new WeakMap;
const myArray = [];
const myObj = {};
```

It would be nice to lint error for those cases when `noExplicitAny:true` is used.

## 🔍 Search Terms

* `noExplicitAny`
* TypeScript compiler API
* `./tsconfig.json`
* `allowJs`
* `checkJs`

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

**Comment by lillallol** [comment]
Also it would be nice to have an example on how to enforce that with the compiler API.

**Comment by MartinJohns** [comment]
https://typescript-eslint.io/rules/no-explicit-any/

**Comment by lillallol** [comment]
> This rule reports on explicit uses of the any keyword as a type annotation.

I do not think this is what I am looking for, since in `.js` you can not use the `any`  keyword.

I also tried `npx eslint ./test.js` and it does not lint error for `any`:

* `./test.js`

```js
const mySet = new Set;
const myMap = new Map;
const myWeakMap = new WeakMap;
const myArray = [];
const myObj = {};
```

* `.eslintrc.js`

```js
module.exports = {
    "env": {
        "browser": true,
        "es2021": true
    },
    "extends": [
        "eslint:recommended",
        "plugin:@typescript-eslint/recommended"
    ],
    "overrides": [
        {
            "env": {
                "node": true
            },
            "files": [
                ".eslintrc.{js,cjs}"
            ],
            "parserOptions": {
                "sourceType": "script"
            }
        }
    ],
    "parser": "@typescript-eslint/parser",
    "parserOptions": {
        "ecmaVersion": "latest",
        "sourceType": "module"
    },
    "plugins": [
        "@typescript-eslint"
    ],
    "rules": {
        "@typescript-eslint/no-explicit-any": "error"
    }
}
```

**Comment by guillaumebrunerie** [comment]
I'm not sure how typescript-eslint works with `.js` files, but you can probably use the rules `no-unsafe-argument`, `no-unsafe-assignment`, `no-unsafe-call`, `no-unsafe-member-access`, and `no-unsafe-return`, which make `any` behave basically like `unknown` by forbidding doing anything with a value of type `any`. Note that your original code does not trigger any of those rules, because you are not doing anything with your Sets/Maps of `any`s, so there is nothing unsound at all in this code.

**Comment by RyanCavanaugh** [comment]
I don't think `noExplicitAny` is a good name for the behavior you're looking to restrict. The existing lint rules about not writing the sequence "``any`" in your code are mostly orthogonal.

There are a few different categories to think about there:
 * Generic inference falling back to a default of `any` due to zero candidates
 * Overload selection choosing an overload that returns `any` when non-`any` overloads exist
 * `myArray` and `myObj` are not `any` by any reasonable definition; not sure what you're looking for here
 * Probably others

This has been discussed before but I don't think anyone's come forth with an explicit proposal about what exactly to do.
